src/neo4j_graphrag/experimental/pipeline/kg_builder.py

```diff
@@ -0,0 +1,221 @@
+    SchemaEntity,
+    SchemaRelation,
+)
+from neo4j_graphrag.experimental.components.text_splitters.fixed_size_splitter import (
+    FixedSizeSplitter,
+)
+from neo4j_graphrag.experimental.pipeline.pipeline import Pipeline, PipelineResult
+from neo4j_graphrag.llm.base import LLMInterface
+
+
+class KnowledgeGraphBuilderConfig(BaseModel):
+    llm: LLMInterface
+    driver: neo4j.Driver
+    file_path: Optional[str] = None
+    text: Optional[str] = None
+    entities: list[SchemaEntity] = Field(default_factory=list)
+    relations: list[SchemaRelation] = Field(default_factory=list)
+    potential_schema: list[tuple[str, str, str]] = Field(default_factory=list)
+    pdf_loader: Any = None
+    kg_writer: Any = None
+    text_splitter: Any = None
+    on_error: OnError = OnError.RAISE
+
+    @model_validator(mode="before")
+    def check_input_source(cls, values: dict[str, Any]) -> dict[str, Any]:
+        file_path = values.get("file_path")
+        text = values.get("text")
+        if (file_path is None and text is None) or (
+            file_path is not None and text is not None
+        ):
+            raise ValueError("Exactly one of 'file_path' or 'text' must be provided.")
+        return values
+
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+
+class KnowledgeGraphBuilder:
+    """
+    A class to simplify the process of building a knowledge graph from text documents.
+    It abstracts away the complexity of setting up the pipeline and its components.
+
+    Args:
+        llm (LLMInterface): An instance of an LLM to use for entity and relation extraction.
+        driver (neo4j.Driver): A Neo4j driver instance for database connection.
+        entities (Optional[List[str]]): A list of entity labels as strings.
+        relations (Optional[List[str]]): A list of relation labels as strings.
+        potential_schema (Optional[List[tuple]]): A list of potential schema relationships.
+        from_pdf (bool): Determines whether to include the PdfLoader in the pipeline.
+                         If True, expects `file_path` input in `run` methods.
+                         If False, expects `text` input in `run` methods.
+        text_splitter (Optional[Any]): A text splitter component. Defaults to FixedSizeSplitter().
+        pdf_loader (Optional[Any]): A PDF loader component. Defaults to PdfLoader().
+        kg_writer (Optional[Any]): A knowledge graph writer component. Defaults to Neo4jWriter().
+        on_error (OnError): Error handling strategy. Defaults to OnError.RAISE.
+    """
+
+    def __init__(
+        self,
+        llm: LLMInterface,
+        driver: neo4j.Driver,
+        entities: Optional[List[str]] = None,
+        relations: Optional[List[str]] = None,
+        potential_schema: Optional[List[tuple[str, str, str]]] = None,
+        from_pdf: bool = True,
+        text_splitter: Optional[Any] = None,
+        pdf_loader: Optional[Any] = None,
+        kg_writer: Optional[Any] = None,
```

Comment:
Is there a reason to have only these configurable components?

I can see two warnings:
- If user provides the KG writer, then the driver is not used - do we want to make it optional?
- I'm a bit afraid it might cause some issues when we add the entity resolver, since it also requires the Neo4j driver as input, I don't know exactly but I feel like we can introduce a lot of mess
- I think we need to add the extractor or the extractor prompt here

That being said, I think we can leave it as it is since this is definitely where we want to go (letting users customize the components), we'll have to consider the points above when we dig deeper into this concept.

Reply:
https://github.com/neo4j/neo4j-graphrag-python/pull/165/commits/eaad20aaecbe132b58cd93cd1e4822ab014237ff